src/mapping/query/arithmetic.rs

```diff
@@ -0,0 +1,406 @@
+}
+
+/// If either value is a floating point number type and the other value is an
+/// integer type the integer is "degraded" into a float. This allows us to
+/// perform arithmetic on common values, but if both are integers then their
+/// precision is preserved.
+fn consistent_number_types(mut left: Value, mut right: Value) -> (Value, Value) {
+    if let Value::Float(_) = left {
+        if let Value::Integer(ri) = right {
+            right = Value::Float(ri as f64);
+        }
+    } else if let Value::Float(_) = right {
+        if let Value::Integer(li) = left {
+            left = Value::Float(li as f64);
+        }
+    }
+    (left, right)
+}
+
+// Degrades non-float numerical types into floats for the purposes of convenient
+// boolean comparison.
+fn compare_number_types(
+    left: Value,
+    right: Value,
+    compare_fn: &dyn Fn(f64, f64) -> bool,
+) -> Result<Value> {
+    let (left, right) = consistent_number_types(left, right);
+    Ok(Value::Boolean(match left {
+        Value::Float(fl) => match right {
+            Value::Float(fr) => compare_fn(fl, fr),
+            vr => return Err(format!("unable to compare right-hand field type {:?}", vr)),
+        },
+        Value::Integer(il) => match right {
+            Value::Integer(ir) => compare_fn(il as f64, ir as f64),
+            vr => return Err(format!("unable to compare right-hand field type {:?}", vr)),
+        },
+        vl => return Err(format!("unable to compare left-hand field type {:?}", vl)),
+    }))
+}
+
+impl Function for Arithmetic {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let left = self.left.execute(ctx);
+        let right = self.right.execute(ctx);
+
+        // TODO: A lot of these comparisons could potentially be baked into the
+        // Value type. However, we would need to agree on general rules as to
+        // how different types are compared.
+
+        Ok(match self.op {
+            Operator::Multiply => {
+                let (left, right) = consistent_number_types(left?, right?);
+                match left {
+                    Value::Float(fl) => match right {
+                        Value::Float(fr) => Value::Float(fl * fr),
+                        vr => {
+                            return Err(format!(
+                                "unable to multiply right-hand field type {:?}",
+                                vr
+                            ))
+                        }
+                    },
+                    Value::Integer(il) => match right {
+                        Value::Integer(ir) => Value::Integer(il * ir),
+                        vr => {
+                            return Err(format!(
+                                "unable to multiply right-hand field type {:?}",
+                                vr
+                            ))
+                        }
+                    },
+                    vl => return Err(format!("unable to multiply left-hand field type {:?}", vl)),
```

Comment:
Any consideration for `String * Integer` to produce a repeating string? Similarly, `Array * Integer` to extend an array?

Reply:
See https://github.com/timberio/vector/issues/3728